From bda292f9c2a266923c7a33ed5a10aba687c043e8 Mon Sep 17 00:00:00 2001
From: Praveen Chavan <pchavan@codeaurora.org>
Date: Tue, 21 Jun 2016 18:59:40 -0700
Subject: [PATCH 1/3] mm-video-v4l2: vdec: de-couple c2d output-buf reqs from
 venus

When C2D is enabled (for on-the-fly-color-conversion), output
buffer requirements are dictated by C2D hardware.
Keep these requirements aside and do not override with the
Venus output buffer requirements.

Route updating of buffer requirements via c2d proxy.
Based on whether c2d is enabled, requirements will be handed
to the vidc driver appropriately

Also, allow changing C2D buffer requirements only when
port is being reconfigured to ensure that c2d buffer
sizes reflect the actual allocation sizes.

Change-Id: If64165f19c7210121a205375c0e460cb873a9d92
---
 mm-video-v4l2/vidc/vdec/inc/omx_vdec.h        |  3 +
 mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp | 97 ++++++++++++++++++++++++---
 2 files changed, 89 insertions(+), 11 deletions(-)

diff --git a/mm-video-v4l2/vidc/vdec/inc/omx_vdec.h b/mm-video-v4l2/vidc/vdec/inc/omx_vdec.h
index cb580b6..8730b89 100644
--- a/mm-video-v4l2/vidc/vdec/inc/omx_vdec.h
+++ b/mm-video-v4l2/vidc/vdec/inc/omx_vdec.h
@@ -1003,6 +1003,7 @@ class omx_vdec: public qc_omx_component
                 bool get_color_format(OMX_COLOR_FORMATTYPE &dest_color_format);
                 bool update_buffer_req();
                 bool get_buffer_req(unsigned int &buffer_size);
+                OMX_ERRORTYPE set_buffer_req(OMX_U32 buffer_size, OMX_U32 actual_count);
                 OMX_BUFFERHEADERTYPE* get_il_buf_hdr();
                 OMX_BUFFERHEADERTYPE* get_il_buf_hdr(OMX_BUFFERHEADERTYPE *input_hdr);
                 OMX_BUFFERHEADERTYPE* get_dr_buf_hdr(OMX_BUFFERHEADERTYPE *input_hdr);
@@ -1025,6 +1026,8 @@ class omx_vdec: public qc_omx_component
                 unsigned int allocated_count;
                 unsigned int buffer_size_req;
                 unsigned int buffer_alignment_req;
+                OMX_U32 m_c2d_width;
+                OMX_U32 m_c2d_height;
                 OMX_QCOM_PLATFORM_PRIVATE_LIST      m_platform_list_client[MAX_COUNT];
                 OMX_QCOM_PLATFORM_PRIVATE_ENTRY     m_platform_entry_client[MAX_COUNT];
                 OMX_QCOM_PLATFORM_PRIVATE_PMEM_INFO m_pmem_info_client[MAX_COUNT];
diff --git a/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp b/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
index 243907a..c56f097 100644
--- a/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
+++ b/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
@@ -3825,6 +3825,19 @@ OMX_ERRORTYPE  omx_vdec::set_parameter(OMX_IN OMX_HANDLETYPE     hComp,
                                   #else
                                      if (!port_format_changed) {
                                   #endif
+                                       // Buffer count can change only when port is disabled
+                                       if (!release_output_done()) {
+                                           DEBUG_PRINT_ERROR("Cannot change o/p buffer count since all buffers are not freed yet !");
+                                           eRet = OMX_ErrorInvalidState;
+                                           break;
+                                       }
+                                  #ifdef _COLOR_CONVERTION_
+                                       // route updating of buffer requirements via c2d proxy.
+                                       // Based on whether c2d is enabled, requirements will be handed
+                                       // to the vidc driver appropriately
+                                       eRet = client_buffers.set_buffer_req(portDefn->nBufferSize,
+                                                portDefn->nBufferCountActual);
+                                  #else
                                        if ( portDefn->nBufferCountActual >= drv_ctx.op_buf.mincount &&
                                                portDefn->nBufferSize >=  drv_ctx.op_buf.buffer_size ) {
                                            drv_ctx.op_buf.actualcount = portDefn->nBufferCountActual;
@@ -3833,11 +3846,13 @@ OMX_ERRORTYPE  omx_vdec::set_parameter(OMX_IN OMX_HANDLETYPE     hComp,
                                            drv_ctx.extradata_info.size = drv_ctx.extradata_info.count *
                                                drv_ctx.extradata_info.buffer_size;
                                            eRet = set_buffer_req(&drv_ctx.op_buf);
-                                           if (eRet == OMX_ErrorNone)
-                                               m_port_def = *portDefn;
+                                       }
+                                  #endif
+                                       if (eRet == OMX_ErrorNone) {
+                                           m_port_def = *portDefn;
                                        } else {
                                            DEBUG_PRINT_ERROR("ERROR: OP Requirements(#%d: %u) Requested(#%u: %u)",
-                                                   drv_ctx.op_buf.mincount, (unsigned int)drv_ctx.op_buf.buffer_size,
+                                                   drv_ctx.op_buf.mincount, (unsigned int)buffer_size,
                                                    (unsigned int)portDefn->nBufferCountActual, (unsigned int)portDefn->nBufferSize);
                                            eRet = OMX_ErrorBadParameter;
                                        }
@@ -10895,6 +10910,8 @@ omx_vdec::allocate_color_convert_buf::allocate_color_convert_buf()
     init_members();
     ColorFormat = OMX_COLOR_FormatMax;
     dest_format = YCbCr420P;
+    m_c2d_width = 0;
+    m_c2d_height = 0;
 }
 
 void omx_vdec::allocate_color_convert_buf::set_vdec_client(void *client)
@@ -10907,6 +10924,7 @@ void omx_vdec::allocate_color_convert_buf::init_members()
     allocated_count = 0;
     buffer_size_req = 0;
     buffer_alignment_req = 0;
+    m_c2d_width = m_c2d_height = 0;
     memset(m_platform_list_client,0,sizeof(m_platform_list_client));
     memset(m_platform_entry_client,0,sizeof(m_platform_entry_client));
     memset(m_pmem_info_client,0,sizeof(m_pmem_info_client));
@@ -10937,6 +10955,20 @@ bool omx_vdec::allocate_color_convert_buf::update_buffer_req()
         return status;
     }
     pthread_mutex_lock(&omx->c_lock);
+
+    bool resolution_upgrade = (omx->drv_ctx.video_resolution.frame_height > m_c2d_height ||
+            omx->drv_ctx.video_resolution.frame_width > m_c2d_width);
+    if (resolution_upgrade) {
+        // resolution upgraded ? ensure we are yet to allocate;
+        // failing which, c2d buffers will never be reallocated and bad things will happen
+        if (allocated_count > 0) {
+            DEBUG_PRINT_ERROR("Cannot change C2D buffer requirements with %d active allocations",
+                    allocated_count);
+            status = false;
+            goto fail_update_buf_req;
+        }
+    }
+
     if (omx->drv_ctx.output_format != VDEC_YUV_FORMAT_NV12 &&
             ColorFormat != OMX_COLOR_FormatYUV420Planar) {
         DEBUG_PRINT_ERROR("update_buffer_req: Unsupported color conversion");
@@ -10962,12 +10994,12 @@ bool omx_vdec::allocate_color_convert_buf::update_buffer_req()
             status = false;
             c2d.close();
             buffer_size_req = 0;
+            // TODO: make this fatal. Driver is not supposed to quote size
+            //  smaller than what C2D needs !!
         } else {
             buffer_size_req = destination_size;
-            if (buffer_size_req < omx->drv_ctx.op_buf.buffer_size)
-                buffer_size_req = omx->drv_ctx.op_buf.buffer_size;
-            if (buffer_alignment_req < omx->drv_ctx.op_buf.alignment)
-                buffer_alignment_req = omx->drv_ctx.op_buf.alignment;
+            m_c2d_height = omx->drv_ctx.video_resolution.frame_height;
+            m_c2d_width = omx->drv_ctx.video_resolution.frame_width;
         }
     }
 fail_update_buf_req:
@@ -11118,14 +11150,57 @@ OMX_BUFFERHEADERTYPE* omx_vdec::allocate_color_convert_buf::get_il_buf_hdr()
             goto fail_get_buffer_size;
         }
     }
-    if (buffer_size < omx->drv_ctx.op_buf.buffer_size)
-        buffer_size = omx->drv_ctx.op_buf.buffer_size;
-    if (buffer_alignment_req < omx->drv_ctx.op_buf.alignment)
-        buffer_alignment_req = omx->drv_ctx.op_buf.alignment;
 fail_get_buffer_size:
     pthread_mutex_unlock(&omx->c_lock);
     return status;
 }
+
+OMX_ERRORTYPE omx_vdec::allocate_color_convert_buf::set_buffer_req(
+        OMX_U32 buffer_size, OMX_U32 actual_count) {
+    OMX_U32 expectedSize = enabled ? buffer_size_req : omx->drv_ctx.op_buf.buffer_size;
+
+    if (buffer_size < expectedSize) {
+        DEBUG_PRINT_ERROR("OP Requirements: Client size(%u) insufficient v/s requested(%u)",
+                buffer_size, expectedSize);
+        return OMX_ErrorBadParameter;
+    }
+    if (actual_count < omx->drv_ctx.op_buf.actualcount) {
+        DEBUG_PRINT_ERROR("OP Requirements: Client count(%u) insufficient v/s requested(%u)",
+                actual_count, omx->drv_ctx.op_buf.actualcount);
+        return OMX_ErrorBadParameter;
+    }
+
+    bool reqs_updated = false;
+    if (enabled) {
+        // disallow changing buffer size/count while we have active allocated buffers
+        if (allocated_count > 0) {
+            DEBUG_PRINT_ERROR("Cannot change C2D buffer size from %u to %u with %d active allocations",
+                    buffer_size_req, buffer_size, allocated_count);
+            return OMX_ErrorInvalidState;
+        }
+
+        buffer_size_req = buffer_size;
+    } else {
+        if (buffer_size > omx->drv_ctx.op_buf.buffer_size) {
+            omx->drv_ctx.op_buf.buffer_size = buffer_size;
+            reqs_updated = true;
+        }
+    }
+
+    if (actual_count > omx->drv_ctx.op_buf.actualcount) {
+        omx->drv_ctx.op_buf.actualcount = actual_count;
+        reqs_updated = true;
+    }
+
+    if (reqs_updated) {
+        omx->drv_ctx.extradata_info.count = omx->drv_ctx.op_buf.actualcount;
+        omx->drv_ctx.extradata_info.size = omx->drv_ctx.extradata_info.count *
+                omx->drv_ctx.extradata_info.buffer_size;
+        return omx->set_buffer_req(&(omx->drv_ctx.op_buf));
+    }
+    return OMX_ErrorNone;
+}
+
 OMX_ERRORTYPE omx_vdec::allocate_color_convert_buf::free_output_buffer(
         OMX_BUFFERHEADERTYPE *bufhdr)
 {
-- 
1.9.1

